include/odp/api/spec/crypto.h

```diff
@@ -316,12 +316,35 @@ int odp_crypto_result(odp_crypto_packet_result_t *result,
  * The value of pktout[n] is ignored as pktout[n] is used purely as an
  * output parameter that returns the handle of the newly allocated packet.
  *
+ * ODP_CRYPTO_OP_TYPE_OOP:
+ *
+ * Writes the output bytes of the crypto operation in a caller provided
+ * output packet passed through pkt_out[n]. Input packets are not consumed
+ * nor modified. Memory layout (including packet data pointers, head and
+ * tail room, segmentation) of the output packet may change during the
+ * operation.
+ *
+ * Crypto output is the processed crypto_range, auth_range and
+ * MAC/digest (in encode sessions) of the input packet. The operation
+ * behaves as if crypto range and auth range were first copied from the
+ * input packet to the output packet and then the crypto operation
+ * was applied to the output packet.
```

Comment:
I am afraid there might be slight ambiguity here in a couple cases:

1) If a cipher or auth algorithm is null (the other algorithm may or may not be null), is the corresponding cipher or auth range copied to the packet? The current text implies that it is, and it would be kind-of consistent and maybe also handy at times, but I wonder if we should specify that the range of the null algorithm is ignored for this purpose.

2) Algorithms providing authenticated encryption ignore the auth_range parameter. Maybe we should state that it is ignored also for the purpose of the conceptual pre-crypto data copy in this OOP case.

Reply:
> The current text implies that it is, and it would be kind-of consistent and maybe also handy at times, but I wonder if we should specify that the range of the null algorithm is ignored for this purpose.

There are two cases here,
1. For cipher only cases etc, auth would be NULL and do we want copy to happen based on auth range? I guess we should not do that. And so, it is fair to say that the range is ignored.
2. For NULL cipher + NULL auth, do we need to copy? If yes, based on which range?

> Algorithms providing authenticated encryption ignore the auth_range parameter. Maybe we should state that it is ignored also for the purpose of the conceptual pre-crypto data copy in this OOP case.

Is that the case? At least for PDCP, I'm aware of a case where authenticated range and ciphered range would be different.

Reply:
> For cipher only cases etc, auth would be NULL and do we want copy to happen based on auth range? I guess we should not do that.

Then I think an API update is needed. If the range would be copied even with null algorithm, one could still avoid the copy by setting range length to zero. But I think just ignoring the range might be better.

> For NULL cipher + NULL auth, do we need to copy? If yes, based on which range?

I do not think null+null should be a special case. If null range is ignored when the other algorithm is not null, then in this case both ranges would be ignored and nothing would get copied. OTOH, if null range is not ignored, then in this null-null case both ranges would get copied.

I think I will update the API a bit regarding this after discussing internally. Maybe I also add clarification on how hash_result_offset is interpreted in the case of OOP with a shift (shift is not applied, but whether the offset applies to the input packet or output packet depends on the direction (encode vs. decode).

Reply:
Agreed.